lib/Transforms/YosysOptimizer/YosysOptimizer.cpp

```diff
@@ -0,0 +1,122 @@
+
+  void runOnOperation() override;
+
+  llvm::StringRef getArgument() const override { return "yosys-optimizer"; }
+
+  void getDependentDialects(mlir::DialectRegistry &registry) const override {
+    registry.insert<mlir::arith::ArithDialect, mlir::heir::comb::CombDialect>();
+  }
+
+ private:
+  // Path to a directory containing yosys techlibs.
+  std::string runfiles;
+  // Path to ABC binary.
+  std::string abcPath;
+
+  bool abcFast;
+};
+
+// Globally optimize an MLIR module.
+void YosysOptimizer::runOnOperation() {
+  getOperation().walk([&](func::FuncOp op) {
+    // Translate function to Verilog. Translation will fail if the func
+    // contains unsupported operations.
+    // FIXME: Directly convert MLIR to Yosys' AST instead of using Verilog.
+    char *filename = tmpnam(NULL);
+    std::error_code EC;
+    llvm::raw_fd_ostream of(filename, EC);
+    if (failed(translateToVerilog(op, of)) || EC) {
+      return WalkResult::interrupt();
+    }
+    of.close();
+
+    // Invoke Yosys to translate to a combinational circuit and optimize.
+    Yosys::yosys_setup();
+    Yosys::log_error_stderr = true;
+    Yosys::run_pass(llvm::formatv(kYosysTemplate.data(), filename,
+                                  op.getSymName(), runfiles, abcPath,
+                                  abcFast ? "-fast" : ""));
+
+    // Translate to MLIR and insert into the func
+    std::stringstream cellOrder;
+    Yosys::log_streams.push_back(&cellOrder);
+    Yosys::run_pass("torder -stop * P*;");
+    Yosys::log_streams.clear();
+    auto topologicalOrder = getTopologicalOrder(cellOrder);
+
+    // Insert the optimized MLIR.
+    LUTImporter lutImporter = LUTImporter(&getContext());
+    Yosys::RTLIL::Design *design = Yosys::yosys_get_design();
+    func::FuncOp func =
+        lutImporter.importModule(design->top_module(), topologicalOrder);
+    op.getBody().takeBody(func.getBody());
+
+    func.dump();
+
+    return WalkResult::advance();
+  });
+  Yosys::yosys_shutdown();
+}
+
+std::unique_ptr<mlir::Pass> createYosysOptimizer(std::string runfiles,
+                                                 std::string abcPath,
+                                                 bool abcFast) {
+  return std::make_unique<YosysOptimizer>(runfiles, abcPath, abcFast);
+}
+
+std::function<void(OpPassManager &pm,
+                   const YosysOptimizerPipelineOptions &options)>
+buildYosysPipelineLambda(std::string runfiles, std::string abcPath) {
+  return [runfiles, abcPath](OpPassManager &pm,
+                             const YosysOptimizerPipelineOptions &options) {
+    pm.addPass(createYosysOptimizer(runfiles, abcPath, options.abcFast));
+    pm.addPass(mlir::createCanonicalizerPass());
+  };
+}
+
+void registerYosysOptimizerPipeline(std::string runfiles, std::string abcPath) {
+  PassPipelineRegistration<YosysOptimizerPipelineOptions>(
+      "yosys-optimizer", "The yosys optimizer pipeline.",
+      buildYosysPipelineLambda(runfiles, abcPath));
+}
```

Comment:
Would the following simplify the lambda? I'm just not quite sure why it needs the extra function to generate the lambda.

```
void registerYosysOptimizerPipeline(std::string runfiles, std::string abcPath) {
  PassPipelineRegistration<YosysOptimizerPipelineOptions>(
      "yosys-optimizer", "The yosys optimizer pipeline.",
       [&](OpPassManager &pm, const YosysOptimizerPipelineOptions &options) {
         pm.addPass(createYosysOptimizer(runfiles, abcPath, options.abcFast));
         pm.addPass(mlir::createCanonicalizerPass());
       }));
}
```